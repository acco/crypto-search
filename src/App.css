.CryptoTable {
  margin-top: 15vh;
  width: 100%;
}

.loading-ring {
  text-align: center;
}
