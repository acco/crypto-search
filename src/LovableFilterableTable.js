import React, { Component } from "react";

import "./LovableFilterableTable.css";

const isLovedStyle = {
  backgroundColor: "rgba(255, 182, 193, 0.02)"
};

const tdStyle = {
  padding: "5px"
};

const getObjectValues = item => {
  // I love you, JavaScript!!!
  if (typeof Object.values === "function") {
    return Object.values(item);
  } else {
    return Object.keys(item).map(k => item[k]);
  }
};

const filterMatches = (filter, items) => {
  return items.filter(item => {
    return getObjectValues(item).find(
      property => property && property.match(new RegExp(filter, "i"))
    );
  });
};

class LovableFilterableTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      filter: "",
      matches: this.props.items
    };
  }

  updateFilter = (filter, items) => {
    this.setState(() => ({
      filter: filter,
      matches: filterMatches(filter, items)
    }));
  };

  onFilterChange = e => {
    this.updateFilter(e.target.value, this.props.items);
  };

  componentWillReceiveProps(nextProps) {
    this.updateFilter(this.state.filter, nextProps.items);
  }

  render() {
    const { schema, onHeartClick } = this.props;
    const keys = Object.keys(schema);
    const { filter, matches } = this.state;

    return (
      <div className="LovableFilterableTable">
        <div className="row">
          <table>
            <thead>
              <tr>
                <td>
                  <label htmlFor="filterField">Filter</label>
                  <input
                    type="text"
                    placeholder="Filter..."
                    id="filterField"
                    style={{ minWidth: "300px" }}
                    onChange={this.onFilterChange}
                    value={filter}
                  />
                </td>
              </tr>
              <tr>
                {keys.map(key =>
                  <th key={key}>
                    {schema[key]}
                  </th>
                )}
                <th />
              </tr>
            </thead>
            <tbody>
              {matches.map(item => {
                const heartClass = item.isLoved
                  ? "fa fa-heart"
                  : "fa fa-heart-o";
                return (
                  <tr style={item.isLoved ? isLovedStyle : {}} key={item.id}>
                    {keys.map(key => {
                      const value = item[key];
                      return (
                        <td key={key} className={`item-${key}`} style={tdStyle}>
                          {value}
                        </td>
                      );
                    })}
                    <td className="item-heart">
                      {
                        <i
                          className={heartClass}
                          aria-hidden="true"
                          onClick={() => onHeartClick(item.id)}
                        />
                      }
                    </td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

export default LovableFilterableTable;
