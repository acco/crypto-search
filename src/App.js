import React, { Component } from "react";
import "./App.css";
import loadingRing from "./loading-ring.svg";

import LovableFilterableTable from "./LovableFilterableTable";

export const tableSchema = {
  name: "Name",
  symbol: "Symbol",
  price_usd: "Price ($)",
  percent_change_1h: "% Change (1h)",
  percent_change_24h: "% Change (24h)",
  percent_change_7d: "% Change (7d)"
};

const API_URL = "https://api.coinmarketcap.com/v1/ticker/?limit=100";

const parseJSON = resp => resp.json();

// const extractCoinData = json => json.Data;

// const byNameToArray = byName => {
//   return Object.keys(byName).reduce((memo, name) => {
//     memo.push(byName[name]);
//     return memo;
//   }, []);
// };

// const top100BySortOrder = coins => {
//   return coins
//     .sort((a, b) => parseInt(a.SortOrder, 10) - parseInt(b.SortOrder, 10))
//     .slice(0, 99);
// };

// const snakeCaseKeys = array => {
//   return array.map(o => {
//     return Object.keys(o).reduce((memo, key) => {
//       const newKey = key.replace(/^(\w)/, match => match.toLowerCase());
//       memo[newKey] = o[key];
//       return memo;
//     }, {});
//   });
// };

const ContentWithLoadingIndicator = ({ loaded, children }) => {
  if (loaded) {
    return children;
  } else {
    return (
      <div className="loading-ring">
        <img src={loadingRing} />
      </div>
    );
  }
};

class App extends Component {
  state = {
    coins: [],
    loves: [],
    loaded: false
  };

  handleHeartClick = id => {
    this.setState(prevState => {
      const prevLoves = prevState.loves;
      if (prevLoves.find(i => i === id)) {
        return {
          loves: prevLoves.filter(i => i !== id)
        };
      } else {
        return {
          loves: prevLoves.concat(id)
        };
      }
    });
  };

  fetchCoins = () => {
    fetch(API_URL).then(parseJSON).then(top100 =>
      this.setState({
        coins: top100,
        loaded: true
      })
    );
  };

  componentDidMount() {
    this.fetchCoins();
    setInterval(this.fetchCoins, 10000);
  }

  render() {
    const { loves, loaded } = this.state;

    const coins = this.state.coins.map(coin => ({
      ...coin,
      isLoved: loves.find(id => id === coin.id)
    }));

    return (
      <div className="container">
        <div className="CryptoTable">
          <ContentWithLoadingIndicator loaded={loaded}>
            <LovableFilterableTable
              schema={tableSchema}
              items={coins}
              onHeartClick={this.handleHeartClick}
            />
          </ContentWithLoadingIndicator>
        </div>
      </div>
    );
  }
}

export default App;

// const localStorageKey = "crypto-search:loves";

// class App extends Component {
//   constructor(props) {
//     super(props);

//     this.state = {
//       coins: [],
//       loves: [],
//       loaded: false
//     };

//     if (window && window.localStorage) {
//       const stored = window.localStorage.getItem(localStorageKey);
//       if (stored) {
//         this.state.loves = JSON.parse(stored);
//       }
//     }
//   }

//   storeLoveInLocalStorage(id) {
//     if (window && window.localStorage) {
//       let json = [];

//       const stored = window.localStorage.getItem(localStorageKey);
//       if (stored) {
//         json = JSON.parse(stored);
//       }

//       json.push(id);
//       window.localStorage.setItem(localStorageKey, JSON.stringify(json));
//     }
//   }

//   handleHeartClick = id => {
//     this.setState(prevState => {
//       const prevLoves = prevState.loves;
//       if (prevLoves.find(i => i === id)) {
//         return {
//           loves: prevLoves.filter(i => i !== id)
//         };
//       } else {
//         return {
//           loves: prevLoves.concat(id)
//         };
//       }
//     });
//   };

//   fetchCoins = () => {
//     fetch(API_URL).then(parseJSON).then(top100 =>
//       this.setState({
//         coins: top100,
//         loaded: true
//       })
//     );
//   };

//   componentDidMount() {
//     this.fetchCoins();
//     setInterval(this.fetchCoins, 10000);
//   }

//   render() {
//     const { loves, loaded } = this.state;

//     const coins = this.state.coins.map(coin => ({
//       ...coin,
//       isLoved: loves.find(id => id === coin.id)
//     }));

//     return (
//       <div className="container">
//         <div className="CryptoTable">
//           <ContentWithLoadingIndicator loaded={loaded}>
//             <LovableFilterableTable
//               schema={tableSchema}
//               items={coins}
//               onHeartClick={this.handleHeartClick}
//             />
//           </ContentWithLoadingIndicator>
//         </div>
//       </div>
//     );
//   }
// }

// export default App;
