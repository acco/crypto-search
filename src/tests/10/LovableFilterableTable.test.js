import React from "react";
import ReactDOM from "react-dom";
import LovableFilterableTable from "../../LovableFilterableTable";
import { shallow } from "enzyme";

import { tableSchema } from "../App";

import fs from "fs";
import path from "path";

const SAMPLE_RESPONSE_FILE = path.join(__dirname, "../../sample-data.json");

const generateItems = (n = 30) => {
  const response = fs.readFileSync(SAMPLE_RESPONSE_FILE);
  const json = JSON.parse(response);

  return json.slice(0, n).map(item => ({
    ...item,
    isLoved: false
  }));
};

import serializer from "jest-serializer-enzyme";
expect.addSnapshotSerializer(serializer);

describe("LovableFilterableTable", () => {
  let wrapper;

  it("renders without crashing", () => {
    const items = [];

    const div = document.createElement("div");
    ReactDOM.render(
      <LovableFilterableTable items={items} schema={tableSchema} />,
      div
    );
  });

  describe("when given empty `items`", () => {
    const items = [];

    beforeEach(() => {
      wrapper = shallow(
        <LovableFilterableTable items={[]} schema={tableSchema} />
      );
    });

    it("should render an empty table", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when given some `items`", () => {
    beforeEach(() => {
      const items = generateItems(3);
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      );
    });

    it("should render each item in the table", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("user enters search query", () => {
    let items;

    beforeEach(() => {
      items = generateItems();
      wrapper = shallow(
        <LovableFilterableTable items={items} schema={tableSchema} />
      );

      const searchBox = wrapper.find("input");
      searchBox.simulate("change", { target: { value: "coin" } });
    });

    it("should filter items", () => {
      expect(
        wrapper.find("tbody > tr > .item-name").map(n => n.text())
      ).toMatchSnapshot();
    });

    describe("user clears search query", () => {
      beforeEach(() => {
        const searchBox = wrapper.find("input");
        searchBox.simulate("change", { target: { value: "" } });
      });

      it("should render all the items again", () => {
        expect(wrapper.find("tbody > tr").length).toEqual(items.length);
      });
    });
  });
});
