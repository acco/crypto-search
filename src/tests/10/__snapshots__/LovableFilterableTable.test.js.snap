// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LovableFilterableTable user enters search query should filter items 1`] = `
Array [
  "Bitcoin",
  "Litecoin",
  "Siacoin",
  "Bytecoin",
  "Dogecoin",
  "MaidSafeCoin",
]
`;

exports[`LovableFilterableTable when given empty \`items\` should render an empty table 1`] = `
<div
  className="LovableFilterableTable"
>
  <div
    className="row"
  >
    <table>
      <thead>
        <tr>
          <td>
            <label
              htmlFor="filterField"
            >
              Filter
            </label>
            <input
              id="filterField"
              onChange={[Function]}
              placeholder="Filter..."
              style={
                Object {
                  "minWidth": "300px",
                }
              }
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <th>
            Name
          </th>
          <th>
            Symbol
          </th>
          <th>
            Price ($)
          </th>
          <th>
            % Change (1h)
          </th>
          <th>
            % Change (24h)
          </th>
          <th>
            % Change (7d)
          </th>
          <th />
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`LovableFilterableTable when given some \`items\` should render each item in the table 1`] = `
<div
  className="LovableFilterableTable"
>
  <div
    className="row"
  >
    <table>
      <thead>
        <tr>
          <td>
            <label
              htmlFor="filterField"
            >
              Filter
            </label>
            <input
              id="filterField"
              onChange={[Function]}
              placeholder="Filter..."
              style={
                Object {
                  "minWidth": "300px",
                }
              }
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <th>
            Name
          </th>
          <th>
            Symbol
          </th>
          <th>
            Price ($)
          </th>
          <th>
            % Change (1h)
          </th>
          <th>
            % Change (24h)
          </th>
          <th>
            % Change (7d)
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr
          style={Object {}}
        >
          <td
            className="item-name"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            Bitcoin
          </td>
          <td
            className="item-symbol"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            BTC
          </td>
          <td
            className="item-price_usd"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            2717.71
          </td>
          <td
            className="item-percent_change_1h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            0.13
          </td>
          <td
            className="item-percent_change_24h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            0.17
          </td>
          <td
            className="item-percent_change_7d"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            17.38
          </td>
          <td
            className="item-heart"
          >
            <i
              aria-hidden="true"
              className="fa fa-heart-o"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          style={Object {}}
        >
          <td
            className="item-name"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            Ethereum
          </td>
          <td
            className="item-symbol"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            ETH
          </td>
          <td
            className="item-price_usd"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            339.726
          </td>
          <td
            className="item-percent_change_1h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            0.13
          </td>
          <td
            className="item-percent_change_24h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            1.08
          </td>
          <td
            className="item-percent_change_7d"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            4.95
          </td>
          <td
            className="item-heart"
          >
            <i
              aria-hidden="true"
              className="fa fa-heart-o"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          style={Object {}}
        >
          <td
            className="item-name"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            Ripple
          </td>
          <td
            className="item-symbol"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            XRP
          </td>
          <td
            className="item-price_usd"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            0.299428
          </td>
          <td
            className="item-percent_change_1h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            0.38
          </td>
          <td
            className="item-percent_change_24h"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            -3.25
          </td>
          <td
            className="item-percent_change_7d"
            style={
              Object {
                "padding": "5px",
              }
            }
          >
            18.88
          </td>
          <td
            className="item-heart"
          >
            <i
              aria-hidden="true"
              className="fa fa-heart-o"
              onClick={[Function]}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
